zmq/backend/cython/socket.pyx

```diff
@@ -208,30 +208,109 @@ cdef inline int _proxy_step(void *handle_from, void *handle_to, int max_loops) e
         _raise_errno(errno)
     return 0
 
+cdef struct zmq_part_data_t:
+    zmq_msg_t frame
+    char *msg_c
+    Py_ssize_t msg_c_len
+
+cdef inline void zmq_part_data_init(zmq_part_data_t *part):
+    part[0].msg_c = NULL
+    part[0].msg_c_len = 0
+
+cdef inline int zmq_part_data_asbuffer_r(zmq_part_data_t *part, object msg) except -1:
+    asbuffer_r(msg, <void **>&part[0].msg_c, &part[0].msg_c_len)
+    return 0
+
 cdef inline object _send_copy(void *handle, object msg, int flags=0):
     """Send a message on this socket by copying its content."""
     cdef int rc, rc2
-    cdef zmq_msg_t data
-    cdef char *msg_c
-    cdef Py_ssize_t msg_c_len=0
+    cdef zmq_part_data_t data
 
     # copy to c array:
-    asbuffer_r(msg, <void **>&msg_c, &msg_c_len)
+    zmq_part_data_init(&data)
+    zmq_part_data_asbuffer_r(&data, msg)
 
     # Copy the msg before sending. This avoids any complications with
     # the GIL, etc.
     # If zmq_msg_init_* fails we must not call zmq_msg_close (Bus Error)
-    rc = zmq_msg_init_size(&data, msg_c_len)
+    rc = zmq_msg_init_size(&data.frame, data.msg_c_len)
 
     _check_rc(rc)
 
     with nogil:
-        memcpy(zmq_msg_data(&data), msg_c, zmq_msg_size(&data))
-        rc = zmq_msg_send(&data, handle, flags)
-        rc2 = zmq_msg_close(&data)
+        memcpy(zmq_msg_data(&data.frame), data.msg_c, zmq_msg_size(&data.frame))
+        rc = zmq_msg_send(&data.frame, handle, flags)
+        rc2 = zmq_msg_close(&data.frame)
     _check_rc(rc)
     _check_rc(rc2)
 
+cdef inline object _send_multipart_copy(void *handle, object msg_parts, int flags=0):
+    """Send a multipart message on this socket by copying its content."""
+    cdef int rc, rc2, rc3
+    cdef Py_ssize_t num_parts, base_part, batch_parts, partno
+    cdef int part_flags
+    cdef bint last_batch
+    cdef zmq_part_data_t data[32]
+    cdef object obj
+
+    num_parts = len(msg_parts)
+    base_part = 0
+
+    while base_part < num_parts:
+        # copy to c array:
+        batch_parts = num_parts - base_part
+        if batch_parts > 32:
+            batch_parts = 32
+            last_batch = False
+        else:
+            last_batch = True
+
+        # Copy the msg before sending. This avoids any complications with
+        # the GIL, etc.
+        # If zmq_msg_init_* fails we must not call zmq_msg_close (Bus Error)
+
+        for partno in range(batch_parts):
+            obj = msg_parts[base_part + partno]
+            if isinstance(obj, Frame):
+                obj = obj.buffer
+            zmq_part_data_init(&data[partno])
+            zmq_part_data_asbuffer_r(&data[partno], obj)
+
+        for partno in range(batch_parts):
+            rc = zmq_msg_init_size(&data[partno].frame, data[partno].msg_c_len)
```

Comment:
Why not fuse this loop with the one above ? I've seen gcc be smart enough to do it on its own, but I doubt it would be able here given cython's propensity at generating garbage.

Reply:
If init_size fails (it can), we want to have the whole array (or the part we need) properly initialized to avoid issues doing the cleanup.

Reply:
Not sure I follow. What I propose is something like:

```
  zmq_part_data_init(&data[partno])
  zmq_part_data_asbuffer_r(&data[partno], obj)
  rc = zmq_msg_init_size(&data[partno].frame, data[partno].msg_c_len)
  if rc < 0:
    for partno in range(partno):
      zmq_msg_close(&data[partno].frame)
    break
```

instead of 2 separate loops.

Reply:
Since `zmq_part_data_asbuffer_r` can fail on its own, it gets more complicated than that. Separate loops are simpler re fault handling.